### ISSUE TITLE
Frequent "node marked dead" errors after accumulating many succeeded tasks and dead actors in Ray Train

### ISSUE THREAD

**Issue by Hxinyue** [ray-project/ray#59193]
### What happened + What you expected to happen

1、I'm encountering repeated warnings that nodes are being marked as dead due to missed heartbeats, even though the underlying infrastructure appears healthy:

The node with node id: 5d0da9d7f80164c7d61df8addb6a5d80a9d36ec3d8f988c0eef6b545 and address: 192.168.10.175 and node name: 192.168.10.175 has been marked dead because the detector has missed too many heartbeats from it. This can happen when a 	(1) raylet crashes unexpectedly (OOM, etc.) 
	(2) raylet has lagging heartbeats due to slow network or busy workload.

<img width="1831" height="749" alt="Image" src="https://github.com/user-attachments/assets/0e2d27d1-a63d-4e78-a434-c2dfa279134d" />

Environment & Reproduction:

Network has been verified to be stable with no latency or packet loss.
The issue occurs in the following usage pattern:
Using ray.train to run a same training function (train_func) every 5 minutes.
Each invocation spawns a new worker on the Ray cluster to execute train_func.
Initially, everything runs smoothly for many iterations (hundreds of successful runs).
However, after the cluster accumulates significant metadata—e.g., hundreds of SUCCEEDED tasks, thousands of actors, and hundreds of dead nodes—the "node marked dead" error starts appearing.
Observed Behavior:

During job startup, Ray appears to attempt reconnecting to old IP addresses (e.g., 192.168.10.175) that belonged to workers from early runs and have long since been terminated and marked as dead.
This suggests that internal metadata (e.g., dead node or actor records) is not being properly cleaned up and may interfere with heartbeat monitoring or node discovery.
Mitigation Attempts:

I suspected that excessive caching of destroyed actors was contributing to the issue, so I set:
`RAY_maximum_gcs_destroyed_actor_cached_count=1000
RAY_DASHBOARD_MAX_ACTORS_TO_CACHE=1000`

<img width="987" height="279" alt="Image" src="https://github.com/user-attachments/assets/fceed495-a424-4a81-ab11-64ed309a26b4" />

However, the problem persists even with these limits in place.

Expected Behavior:

Ray should gracefully handle frequent startup/shutdown of workers via ray.train without accumulating stale state that disrupts cluster health detection.
The heartbeat detector should not be affected by historical dead nodes or large numbers of completed/destroyed actors.
Dead node metadata and destroyed actor records should be evicted or ignored appropriately to avoid false positives in node liveness checks.

### Versions / Dependencies

kuberay-operator: v1.4.0
ray: 2.40.0
python: 3.10

### Reproduction script

from ray.train.torch import TorchTrainer
from ray.train import ScalingConfig

def train_func():
    # PyTorch training code here.
    ...

ray.init(address="ray://ip:port",runtime_env={'working_dir': xxx.py})
scaling_config = ScalingConfig(num_workers=1, use_gpu=True)
trainer = TorchTrainer(train_func, scaling_config=scaling_config)
result = trainer.fit()
ray.shutdown()

### Issue Severity

None

**Comment by edoakes** [comment]
@ZacAttack PTAL

**Comment by Hxinyue** [comment]
@edoakes @edoakes Could you please tell me if there are any new developments regarding this issue? This matter is of great significance to me. Can it be fixed in the ray 2.40.0 version? thank you

**Comment by ZacAttack** [comment]
> [@edoakes](https://github.com/edoakes) [@edoakes](https://github.com/edoakes) Could you please tell me if there are any new developments regarding this issue? This matter is of great significance to me. Can it be fixed in the ray 2.40.0 version? thank you

Apologies @Hxinyue , this one slipped my radar.  Reviewing now.

**Comment by ZacAttack** [comment]
So @Hxinyue, we're not currently seeing an especially large spike in this issue in our other environments.  There are some things though that you can look at though to diagnose.  I see you've done some network analysis, unfortunately, that's only part of the puzzle.

The Raylet on the worker node needs to respond to the GCS healthcheck in a timely manner in order to avoid this.  In order to assess raylet health, you should look at the raylets io_context health first of all (ray_io_context_event_loop_lag_ms).  There should be a metric raylet side you can look at.  This is a histogram metric that has an upper bound on the 10s+ bucket.  If anything is hitting that, it's potentially larger then that value.  So you should intuit that any spikes in that range are of an undeterminably long latency.

A long list of nodes, even if dead, shouldn't impact monitoring of the rest of the cluster.  The reason being is that health checking is pretty much done on a single loop.  So even if there are a lot of dead nodes, all this would really impact is the GCS's time between polls.  There wouldn't be an accumulation here that would cause an issue like what you've described.

In addition, the GCS has a metric called health_check_rpc_latency_ms which is the observed latency from the GCS side of heartbeats sent to Raylets.  Check to see if this metric is outstanding.

Next thing, we have a few environment variables you can play with to tweak the heartbeating behavior so as to be more liberal for your environment (particularly if you are seeing high latency regularly).  They are:

// The amount of time to wait before health checking a node on start
RAY_health_check_initial_delay_ms defaults: 5000
/// The interval between two health check.
RAY_health_check_period_ms defaults: 3000
/// The timeout for a health check.
RAY_health_check_timeout_ms defaults: 10000
/// The threshold to consider a node dead.
RAY_health_check_failure_threshold defaults: 5

And finally, there is some documentation (that's under review) w.r.t. resource isolation.  Your Raylet may be getting starved of resources and can't respond to heartbeat requests in a timely manner.  The documentation is [here](https://github.com/ray-project/ray/pull/60183) (under review but still legit) and you can follow the instructions for setting up resource isolation in your environment in order to stabilize things.

Ok, now that that's out of the way.  Next up is the stale state.  What you're observing is pretty bizarre.  How are you tearing the cluster down/closing the job?  The above should only be happening in cases where the GCS is trying to recover.  A graceful tear down should clear everything up.

**Comment by Hxinyue** [comment]
> So [@Hxinyue](https://github.com/Hxinyue), we're not currently seeing an especially large spike in this issue in our other environments. There are some things though that you can look at though to diagnose. I see you've done some network analysis, unfortunately, that's only part of the puzzle.所以，我们目前在其他环境中并没有看到这个问题特别明显的激增。不过还是有一些你可以看来诊断的问题。我看到你做过一些网络分析，不幸的是，这只是谜题的一部分。
> 
> The Raylet on the worker node needs to respond to the GCS healthcheck in a timely manner in order to avoid this. In order to assess raylet health, you should look at the raylets io_context health first of all (ray_io_context_event_loop_lag_ms). There should be a metric raylet side you can look at. This is a histogram metric that has an upper bound on the 10s+ bucket. If anything is hitting that, it's potentially larger then that value. So you should intuit that any spikes in that range are of an undeterminably long latency.工作节点上的 Raylet 需要及时响应 GCS 健康检查，以避免这种情况。要评估光线电池的健康状况，首先应该关注光线 io_context 的健康状况（ray_io_context_event_loop_lag_ms）。应该有公制的光线边可以查看。这是一个直方图度量，其上界位于 10s+桶上。如果有东西达到这个数值，那可能比这个值还大。所以你应该直觉地知道，这个范围内的任何峰值延迟都无法确定地长。
> 
> A long list of nodes, even if dead, shouldn't impact monitoring of the rest of the cluster. The reason being is that health checking is pretty much done on a single loop. So even if there are a lot of dead nodes, all this would really impact is the GCS's time between polls. There wouldn't be an accumulation here that would cause an issue like what you've described.即使节点列表很长，即使已经死掉，也不应该影响对集群其他部分的监控。原因是健康检查基本上是单一循环完成的。所以即使有很多死节点，这也只是 GCS 轮询间隔时间的延长。这里不会有积累物导致你描述的那种问题。
> 
> In addition, the GCS has a metric called health_check_rpc_latency_ms which is the observed latency from the GCS side of heartbeats sent to Raylets. Check to see if this metric is outstanding.此外，GCS 还有一个称为 health_check_rpc_latency_ms 的指标，即从 GCS 端观察到发送给 Raylet 的心跳延迟。检查一下这个指标是否优秀。
> 
> Next thing, we have a few environment variables you can play with to tweak the heartbeating behavior so as to be more liberal for your environment (particularly if you are seeing high latency regularly). They are:接下来，我们有一些环境变量可以调整心跳行为，使环境更宽松（尤其是当你经常遇到高延迟时）。它们是：
> 
> // The amount of time to wait before health checking a node on start启动时检查节点健康所需的等待时间 RAY_health_check_initial_delay_ms defaults: 5000RAY_health_check_initial_delay_ms 默认次数：5000 /// The interval between two health check.两次健康检查之间的间隔。 RAY_health_check_period_ms defaults: 3000RAY_health_check_period_ms 默认值：3000 /// The timeout for a health check.健康检查的暂停。 RAY_health_check_timeout_ms defaults: 10000RAY_health_check_timeout_ms 默认次数：10000 /// The threshold to consider a node dead.这是认为节点死亡的阈值。 RAY_health_check_failure_threshold defaults: 5RAY_health_check_failure_threshold 默认值：5
> 
> And finally, there is some documentation (that's under review) w.r.t. resource isolation. Your Raylet may be getting starved of resources and can't respond to heartbeat requests in a timely manner. The documentation is [here](https://github.com/ray-project/ray/pull/60183) (under review but still legit) and you can follow the instructions for setting up resource isolation in your environment in order to stabilize things.最后，有一些关于资源隔离的文档（正在审查中）。你的 Raylet 可能资源匮乏，无法及时响应心跳请求。文档正在审核中，但仍然有效，你可以按照设置环境资源隔离的说明来稳定环境。
> 
> Ok, now that that's out of the way. Next up is the stale state. What you're observing is pretty bizarre. How are you tearing the cluster down/closing the job? The above should only be happening in cases where the GCS is trying to recover. A graceful tear down should clear everything up.好了，既然说清楚了。接下来是陈旧状态。你观察到的情况相当奇怪。你是怎么拆除集群/完成工作的？上述情况只在 GCS 试图恢复的情况下发生。优雅地拆除一切应该能让一切变得清晰。

1. According to the official documentation of the ray train usage case, after each training task is completed, the ray.shutdown command should be used to terminate the current ray client connection.
2. Provide an update on the current situation of a cluster that is experiencing the "node marked dead" issue: The current cluster is composed of multiple nodes. Observations have shown that whenever this issue occurs, the worker nodes are not on the same node as the head node. I wonder if this has any impact? Although this error does not occur every time the worker and the head are not on the same node, it happens quite frequently. The other services in this cluster do not have such problems.
3. `Liveness:                     exec [bash -c wget --tries 1 -T 2 -q -O- http://localhost:52365/api/local_raylet_healthz | grep success] delay=30s timeout=2s period=5s #success=1 #failure=120
    Readiness:                    exec [bash -c wget --tries 1 -T 2 -q -O- http://localhost:52365/api/local_raylet_healthz | grep success] delay=10s timeout=2s period=5s #success=1 #failure=10`
Observation reveals that the probe check duration is 30 seconds. The cluster we are currently using is deployed using the kuberay and chart packages. Is there any place where this probe duration can be modified and increased?

**Comment by Hxinyue** [comment]
@ZacAttack I wonder if there has been any progress on this issue? It does occur quite frequently. Thank you for your answer.
